src/gdsync.cpp

```diff
@@ -1822,13 +1826,18 @@ struct gds_qp *gds_create_qp(struct ibv_pd *pd, struct ibv_context *context,
                 return NULL;
         }
 
-        gqp = (struct gds_qp*)calloc(1, sizeof(struct gds_qp));
-        if (!gqp) {
+        memset(&exp_qp_attr, 0, sizeof(struct ibv_exp_qp_init_attr));
+        memcpy(&exp_qp_attr, qp_attr, sizeof(gds_qp_init_attr_t));
```

Comment:
can you really bit copy ibv_qp_init_attr onto the ibv_exp_qp_init_attr version? is that guaranteed to work ?
@haggaie can you take a look?

Reply:
This is the definition of struct ibv_qp_attr_attr and its relevant fields:

struct ibv_qp_cap {
    uint32_t        max_send_wr;
    uint32_t        max_recv_wr;
    uint32_t        max_send_sge;
    uint32_t        max_recv_sge;
    uint32_t        max_inline_data;
};

struct ibv_qp_init_attr {
    void               *qp_context;
    struct ibv_cq          *send_cq;
    struct ibv_cq          *recv_cq;
    struct ibv_srq         *srq;
    struct ibv_qp_cap   cap;
    enum ibv_qp_type    qp_type;
    int         sq_sig_all;
    /* Below is needed for backwards compatabile */
    struct ibv_xrc_domain  *xrc_domain;
};

struct ibv_exp_qp_init_attr has one-to-one mapping but with additional fields.

We mainly memcpy the pointers and some shallow structs. In other words, no deep copy seems to be needed. 

Note: We need to use ibv_exp_create_qp to support peer ops. Hence, directly passing ibv_qp_init_attr is not an option.

Reply:
In general the methodology of extending verbs would mean that it would be okay to memcpy, but it seems like in this case there's some incompatibility. The ibv_qp_init_attr of MLNX OFED includes the xrc_domain pointer for some reason where the comp_mask field would be in the exp version. I think the upstream version doesn't.

Reply:
I can copy field by field and leave out xrc_domain. However, if there is a change in the struct definition such as one more field is introduced, we may need to revisit this part of libgdsync again.

@haggaie Can you suggest a way to do this correctly such that it can withstand future changes in the struct definition?

Reply:
Normally when extending these structs people try to preserve both ABI and source backward compatibility. Also, I don't think ibv_qp_init_attr can grow since it doesn't have a length field or a comp_mask field that would tell the callee which version of the struct it got. So I think it would be fine to copy field by field, or to use memcpy without the xrc_domain field.

Reply:
@haggaie thanks